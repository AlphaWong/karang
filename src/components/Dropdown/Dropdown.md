#### Usage:

```js static
import { Dropdown } from 'lalamove-ui';

// <Dropdown items={[object]} />
```

**Basic**

```js
const items = [
  {
    value: 'hello',
    label: 'Hello world!',
  },
  {
    value: 'bye',
    label: 'Bye!',
  },
];
<Dropdown items={items} defaultLabel="Options available" />
```

**Basic with icon and onChange()**

Example for using `<Dropdown />` with custom `onChange()` function

```js
const Icon = require('components/Icon').default;
const items = [
  {
    icon: <Icon type="pinFill" />,
    value: 'TH_BKK',
    label: 'Bangkok',
  },
  {
    icon: <Icon type="pinFill" />,
    value: 'TH_CNX',
    label: 'Chiang Mai',
  },
];
initialState = { selected: items[0] };
<Dropdown
  items={items}
  selectedItem={state.selected}
  onChange={selection => setState({ selected: selection })}
/>
```

**Cascading menu with icon**

```js
const Icon = require('components/Icon').default;
const items = [
  {
    icon: <Icon type="pinFill" />,
    value: 'TH_BKK',
    label: 'Bangkok',
    options: [{
        value: 'Siam',
        label: 'Siam',
      }, {
        value: 'Phrom_Phong',
        label: 'Phrom Phong',
    }]
  },
  {
    icon: <Icon type="pinFill" />,
    value: 'TH_CNX',
    label: 'Chiang Mai',
  },
];
<Dropdown items={items} />
```
