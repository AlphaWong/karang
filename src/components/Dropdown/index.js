import React, { Component } from 'react';
import {
  string,
  arrayOf,
  shape,
  func,
  bool,
  oneOfType,
  object,
} from 'prop-types';
import Downshift from 'downshift';
import styled from 'styled-components';

import DropdownButton from './components/DropdownButton';
import DropdownList from './components/DropdownList';
import noop from 'utils/noop';

import flag from 'assets/testing.png'; // TODO: remove it later, just for testing

const FlagIcon = styled.img`
  max-width: 24px;
  max-height: 24px;
  vertical-align: middle;
`;

// TODO: Remove it later, just for testing
const items = [
  {
    value: 'Option 1',
    label: 'Option 1',
    icon: <FlagIcon src={flag} />,
    options: [
      {
        value: 'Option 1A',
        label: 'Option 1A',
        icon: <FlagIcon src={flag} />,
      },
      {
        value: 'Option 1B',
        label: 'Option 1B',
        icon: <FlagIcon src={flag} />,
        options: [
          {
            value: 'Option 2A',
            label: 'Option 2A',
            icon: <FlagIcon src={flag} />,
          },
          {
            value: 'Option 2B',
            label: 'Option 2B',
            icon: <FlagIcon src={flag} />,
          },
          {
            value: 'Option 2C',
            label: 'Option 2C',
            icon: <FlagIcon src={flag} />,
          },
        ],
      },
      {
        value: 'Option 1C',
        label: 'Option 1C',
        icon: <FlagIcon src={flag} />,
      },
    ],
  },
  {
    value: 'Option 2',
    label: 'Option 2',
    icon: <FlagIcon src={flag} />,
  },
];

class Dropdown extends Component {
  static propTypes = {
    onChange: func,
  };

  static defaultProps = {
    onChange: noop,
  };

  state = {
    selectedOption: null,
  };

  handleChange = selectedOption => {
    this.setState({ selectedOption });
    this.props.onChange(selectedOption);
  };

  stateReducer = (state, changes) => {
    switch (changes.type) {
      case Downshift.stateChangeTypes.keyDownEnter:
      case Downshift.stateChangeTypes.clickItem:
        if (changes.selectedItem.options) return state;
        return changes;
      default:
        return changes;
    }
  };

  render() {
    const { selectedOption } = this.state;
    return (
      <Downshift
        onChange={this.handleChange}
        itemToString={item => (item ? item.value : '')}
        stateReducer={this.stateReducer}
        // TODO: pass remain props here
      >
        {({
          getInputProps,
          getItemProps,
          getLabelProps,
          getMenuProps,
          getToggleButtonProps,
          isOpen,
          inputValue,
          highlightedIndex,
          selectedItem,
        }) => (
          <div>
            <DropdownButton
              icon={selectedOption && selectedOption.icon}
              label={selectedOption ? selectedOption.label : 'testing'}
              {...getToggleButtonProps()}
            />
            {console.log(highlightedIndex)}
            {console.log(selectedItem)}
            {isOpen && (
              <DropdownList
                items={items}
                highlightedIndex={highlightedIndex}
                getItemProps={getItemProps}
              />
            )}
          </div>
        )}
      </Downshift>
    );
  }
}

export default Dropdown;
