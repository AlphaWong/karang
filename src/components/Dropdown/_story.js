import React, { Component } from 'react';
import { storiesOf } from '@storybook/react';

import Dropdown from './index';
import Icon from '../Icon';

const items = [
  {
    value: 'hello',
    label: 'Hello world!',
  },
  {
    value: 'bye',
    label: 'Bye!',
  },
];

const itemsWithSubmenu = [
  {
    value: 'TH',
    label: 'Thailand',
    icon: <Icon type="pinFill" />,
    options: [
      {
        value: 'TH_BKK',
        label: 'Bangkok',
        icon: <Icon type="pinFill" />,
      },
      {
        value: 'TH_CNX',
        label: 'Chiang Mai',
        icon: <Icon type="pinFill" />,
      },
    ],
  },
  {
    value: 'HK_HKG',
    label: 'Hong Kong',
    icon: <Icon type="pinFill" />,
  },
  {
    value: 'VN_HCM',
    label: 'Vietnam',
    icon: <Icon type="pinFill" />,
  },
  {
    value: 'PH',
    label: 'Philippines',
    icon: <Icon type="pinFill" />,
    options: [
      {
        value: 'PH_MNL',
        label: 'Manila',
        icon: <Icon type="pinFill" />,
      },
      {
        value: 'PH_CEB',
        label: 'Cebu',
        icon: <Icon type="pinFill" />,
      },
    ],
  },
  {
    value: 'SG_SGN',
    label: 'Singapore',
    icon: <Icon type="pinFill" />,
  },
  {
    value: 'TW_TPE',
    label: 'Taiwan',
    icon: <Icon type="pinFill" />,
  },
];

class Wrapper extends Component {
  state = {
    dropdown1: itemsWithSubmenu[5],
    dropdown2: itemsWithSubmenu[1],
  };

  handleChange = (name, item) => {
    this.setState({ [name]: item });
  };

  render() {
    const { dropdown1, dropdown2 } = this.state;
    return (
      <div>
        <h4>Basic</h4>
        <Dropdown items={items} />
        <h4>Basic with icon</h4>
        <Dropdown
          items={items.map(item =>
            Object.assign({}, item, { icon: <Icon type="pinFill" /> })
          )}
        />
        <h4>Cascading menu with icon</h4>
        <Dropdown
          items={itemsWithSubmenu}
          onChange={item => this.handleChange('dropdown1', item)}
          selectedItem={dropdown1}
        />
        <h4 style={{ textAlign: 'right' }}>Cascading menu on right side</h4>
        <div style={{ float: 'right' }}>
          <Dropdown
            items={itemsWithSubmenu}
            onChange={item => this.handleChange('dropdown2', item)}
            selectedItem={dropdown2}
            direction="left"
          />
        </div>
      </div>
    );
  }
}

storiesOf('Dropdown', module)
  .add('Basic', () => <Wrapper />)
  .add('block', () => <Dropdown block items={items} />);
