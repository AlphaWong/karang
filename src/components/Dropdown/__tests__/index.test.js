import React from 'react';
import { render, mount, shallow } from 'enzyme';
import Dropdown from '../index';

const data = [
  {
    value: 'value_1',
    label: 'Label 1',
  },
  {
    value: 'value_2',
    label: 'Label 2',
  },
];

describe('Dropdown', () => {
  describe('Snapshots', () => {
    it('render', () => {
      const wrapper = render(<Dropdown items={data} />);
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('Render', () => {
    it('not to throw', () => {
      expect(() => mount(<Dropdown items={data} />)).not.toThrow();
    });
  });

  describe('Functions', () => {});
});
