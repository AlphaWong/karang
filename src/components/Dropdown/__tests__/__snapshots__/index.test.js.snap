// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown Snapshots render 1`] = `
.c2 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  display: block;
  padding: 0.5em 1em;
  border: 1px currentColor solid;
  background-color: transparent;
  border-radius: 2px;
  font-family: "Noto Sans",sans-serif;
  font-size: 14px;
  font-weight: 400;
  line-height: 1;
  text-align: center;
  cursor: pointer;
  outline: 0;
  display: inline-block;
  border-color: rgba(229,229,229,1);
  color: rgba(88,89,91,1);
}

.c2:disabled {
  opacity: 0.3;
  cursor: not-allowed;
}

.c2:hover:enabled,
.c2:focus:enabled {
  box-shadow: 0 2px 4px rgba(0,0,0,0.22);
}

.c2:active:enabled {
  box-shadow: none;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  min-width: 80px;
  padding: 0.5em 1em;
  background-color: rgba(255,255,255,1);
  white-space: nowrap;
}

.c1:focus:enabled {
  box-shadow: none;
}

.c3 {
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  margin-right: 0.5em;
  line-height: 1.6;
  text-align: left;
}

.c4 {
  margin-right: -5px;
}

.c0 {
  position: relative;
  display: inline-block;
}

<div
  aria-expanded="false"
  aria-haspopup="listbox"
  aria-labelledby="downshift-0-label"
  class="c0"
  role="combobox"
>
  <button
    aria-autocomplete="list"
    aria-haspopup="true"
    aria-label="open menu"
    aria-labelledby="downshift-0-label"
    autocomplete="off"
    class="c1 c2"
    data-toggle="true"
    id="downshift-0-input"
    role="button"
    type="button"
    value=""
  >
    <span
      class="c3"
    >
      Options
    </span>
    <span
      class="c4"
    >
      <svg
        style="vertical-align:middle;fill:rgba(180,180,180,1);width:24px;height:24px"
        version="1.1"
        viewBox="0 0 1024 1024"
        xlink="http://www.w3.org/1999/xlink"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g>
          <svg
            viewBox="0 0 40 40"
          >
            <path
              d="m31 16.4q0 0.3-0.2 0.5l-10.4 10.4q-0.3 0.3-0.5 0.3t-0.6-0.3l-10.4-10.4q-0.2-0.2-0.2-0.5t0.2-0.5l1.2-1.1q0.2-0.2 0.5-0.2t0.5 0.2l8.8 8.8 8.7-8.8q0.3-0.2 0.5-0.2t0.6 0.2l1.1 1.1q0.2 0.2 0.2 0.5z"
            >
              
            </path>
          </svg>
        </g>
      </svg>
    </span>
  </button>
</div>
`;
